File: app/src/main/java/app/resketchware/ui/dialogs/ColorPickerDialog.java
package app.resketchware.ui.dialogs;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.ColorUtils;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import app.resketchware.R;
import app.resketchware.ui.fragments.PalettesFragment;
import app.resketchware.ui.fragments.PickerFragment;

public class ColorPickerDialog extends BottomSheetDialogFragment {

    private PalettesFragment palettesFragment;
    private PickerFragment pickerFragment;

    private TabLayout tabLayout;
    private TextView hexValue;
    private TextView title;
    private MaterialCardView colorPreview;
    private ViewPager2 viewPager;

    private int selectedColor = Color.WHITE;
    private PositiveClickListener positiveClickListener;

    public interface PositiveClickListener {
        void invoke(int color);
    }

    public static ColorPickerDialog newInstance(int selectedColor) {
        ColorPickerDialog fragment = new ColorPickerDialog();
        Bundle args = new Bundle();
        args.putInt("selected_color", selectedColor);
        fragment.setArguments(args);
        return fragment;
    }

    public void setOnPositiveClickListener(PositiveClickListener listener) {
        positiveClickListener = listener;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            selectedColor = getArguments().getInt("selected_color", Color.WHITE);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_color_picker, container, false);

        colorPreview = view.findViewById(R.id.color_preview);
        hexValue = view.findViewById(R.id.hex_value);
        tabLayout = view.findViewById(R.id.tab_layout);
        title = view.findViewById(R.id.title);
        viewPager = view.findViewById(R.id.pager);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        View cancelButton = view.findViewById(R.id.cancel);
        View positiveButton = view.findViewById(R.id.ok);

        cancelButton.setOnClickListener(v -> dismissAllowingStateLoss());
        positiveButton.setOnClickListener(v -> {
            if (positiveClickListener != null) {
                positiveClickListener.invoke(selectedColor);
            }
            dismissAllowingStateLoss();
        });

        ColorPickerPagerAdapter adapter = new ColorPickerPagerAdapter(requireActivity());
        viewPager.setAdapter(adapter);
        viewPager.setOffscreenPageLimit(2);

        new TabLayoutMediator(tabLayout, viewPager, (tab, position) -> {
            int[] tabIcons = {R.drawable.rsw_colorize, R.drawable.rsw_palette};
            tab.setIcon(tabIcons[position]);
        }).attach();

        updateColors(selectedColor);
    }

    private Fragment handleOnNavigationItemSelected(int itemId) {
        switch (itemId) {
            case 0:
                return pickerFragment != null ? pickerFragment : initFragmentAt(itemId);
            default:
                return palettesFragment != null ? palettesFragment : initFragmentAt(itemId);
        }
    }

    private Fragment initFragmentAt(int itemId) {
        switch (itemId) {
            case 0:
                pickerFragment = PickerFragment.newInstance(selectedColor);
                pickerFragment.setOnPickerUpdateListener((r, g, b) -> {
                    int color = Color.rgb(r, g, b);
                    if (selectedColor != color) {
                        selectedColor = color;
                        updateColors(color);
                    }
                });
                return pickerFragment;
            default:
                palettesFragment = PalettesFragment.newInstance();
                palettesFragment.setOnColorSelectListener(color -> {
                    if (selectedColor != color) {
                        selectedColor = color;
                        updateColors(color);
                    }
                });
                return palettesFragment;
        }
    }

    private void updateColors(int color) {
        int contrastColor = ColorUtils.calculateLuminance(color) < 0.35 ? Color.WHITE : Color.BLACK;

        tabLayout.setSelectedTabIndicatorColor(contrastColor);
        tabLayout.setTabIconTint(ColorStateList.valueOf(contrastColor));
        tabLayout.setTabRippleColor(ColorStateList.valueOf(ColorUtils.setAlphaComponent(contrastColor, Math.round(0.25f * 255))));

        hexValue.setText(String.format("#FF%06X", 0xFFFFFF & color));
        hexValue.setTextColor(contrastColor);

        title.setTextColor(contrastColor);

        colorPreview.setCardBackgroundColor(color);

        if (pickerFragment != null) {
            pickerFragment.updateColor(color);
        }
    }

    private class ColorPickerPagerAdapter extends FragmentStateAdapter {

        public ColorPickerPagerAdapter(FragmentActivity fa) {
            super(fa);
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            return handleOnNavigationItemSelected(position);
        }

        @Override
        public int getItemCount() {
            return 2;
        }
    }
}

File: app/src/main/java/app/resketchware/ui/dialogs/ColorPickerDialog.kt
package app.resketchware.ui.dialogs

import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import androidx.core.graphics.ColorUtils
import androidx.core.os.bundleOf
import androidx.core.view.isGone
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2

import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.card.MaterialCardView
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

import app.resketchware.R
import app.resketchware.ui.fragments.PalettesFragment
import app.resketchware.ui.fragments.PickerFragment
import app.resketchware.ui.extensions.toHex
import app.resketchware.ui.extensions.toSurfaceColor

class ColorPickerDialog : BottomSheetDialogFragment() {
  private var palettesFragment: PalettesFragment? = null
  private var pickerFragment: PickerFragment? = null

  private lateinit var tabLayout: TabLayout
  private lateinit var hexValue: TextView
  private lateinit var title: TextView
  private lateinit var colorPreview: MaterialCardView
  private lateinit var viewPager: ViewPager2
  private lateinit var resetButton: View

  private var backupColor = -1
  private var selectedColor = -1
  var onPositiveClick: ((Int) -> Unit)? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    arguments?.getInt("selected_color")?.let { color ->
      backupColor = color
      selectedColor = color
    }
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    val view = inflater.inflate(R.layout.dialog_color_picker, container, false)

    colorPreview = view.findViewById(R.id.color_preview)
    hexValue = view.findViewById(R.id.hex_value)
    tabLayout = view.findViewById(R.id.tab_layout)
    title = view.findViewById(R.id.title)
    viewPager = view.findViewById(R.id.pager)

    return view
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    val cancelButton = view.findViewById<View>(R.id.cancel)
    val positiveButton = view.findViewById<View>(R.id.ok)
    resetButton = view.findViewById(R.id.reset)

    cancelButton.setOnClickListener { dismissAllowingStateLoss() }
    positiveButton.setOnClickListener {
      onPositiveClick?.invoke(selectedColor)
      dismissAllowingStateLoss()
    }
    resetButton.setOnClickListener {
      if (backupColor != selectedColor) {
        it.visibility = View.GONE
        selectedColor = backupColor
        updateColors(selectedColor)
        palettesFragment?.resetPalette()
      }
    }

    val adapter = ColorPickerPagerAdapter(requireActivity())
    viewPager.adapter = adapter
    viewPager.offscreenPageLimit = 2

    TabLayoutMediator(tabLayout, viewPager) { tab, position ->
      val tabIcons = arrayOf(R.drawable.rsw_colorize, R.drawable.rsw_palette)
      tab.setIcon(tabIcons[position])
    }.attach()

    updateColors(selectedColor)
    disableOverScroll()
  }

  override fun onDestroy() {
    super.onDestroy()
    pickerFragment?.let {
      it.onPickerUpdate = null
      it.onResetPalette = null
    }
    palettesFragment?.let {
      it.onColorSelect = null
    }
  }

  private fun handleOnNavigationItemSelected(itemId: Int) = when (itemId) {
    0 -> pickerFragment ?: initFragmentAt(itemId)
    else -> palettesFragment ?: initFragmentAt(itemId)
  }

  private fun initFragmentAt(itemId: Int): Fragment {
    when (itemId) {
      0 -> pickerFragment = PickerFragment.newInstance(selectedColor).apply {
        onPickerUpdate = { r, g, b ->
          val color = Color.rgb(r.toInt(), g.toInt(), b.toInt())
          if (selectedColor != color) {
            updateColors(color)
            selectedColor = color
          }
        }
        onResetPalette = {
          palettesFragment?.resetPalette()
        }
      }
      else -> palettesFragment = PalettesFragment.newInstance().apply {
        onColorSelect = { color ->
          if (selectedColor != color) {
            updateColors(color)
            selectedColor = color
          }
        }
      }
    }
    return handleOnNavigationItemSelected(itemId)
  }

  private fun updateColors(color: Int) {
    val contrastColor = color.toSurfaceColor()

    colorPreview.setCardBackgroundColor(color)

    tabLayout.setSelectedTabIndicatorColor(contrastColor)
    tabLayout.tabIconTint = ColorStateList.valueOf(contrastColor)
    tabLayout.tabRippleColor = ColorStateList.valueOf(
      ColorUtils.setAlphaComponent(contrastColor, 25)
    )

    hexValue.text = color.toHex()
    hexValue.setTextColor(contrastColor)

    title.setTextColor(contrastColor)

    if (resetButton.isGone) {
      resetButton.visibility = View.VISIBLE
    }

    pickerFragment?.updateColor(color)
  }

  private fun disableOverScroll() {
    val view = viewPager.getChildAt(0)
    if (view is RecyclerView) {
      view.overScrollMode = View.OVER_SCROLL_NEVER
    }
  }

  private inner class ColorPickerPagerAdapter(fa: FragmentActivity) : FragmentStateAdapter(fa) {
    override fun createFragment(position: Int): Fragment = handleOnNavigationItemSelected(position)
    override fun getItemCount(): Int = 2
  }

  companion object {
    @JvmStatic
    fun newInstance(selectedColor: Int) = ColorPickerDialog().apply {
      arguments = bundleOf("selected_color" to selectedColor)
    }
  }
}
